---
title: "Using the MethylSeqR package"
author: "Jonathon T. Hill, PhD and Hailey Johnson Zimmerman"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    css: "bioconductor2.css"
  vignette: >
    %\VignetteIndexEntry{Using the MethylSeqR package}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---
```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```
  
```{r, include=FALSE}
library(knitr, quietly=TRUE)
library(MethylSeqR, quietly=TRUE)
# library(Biostrings, quietly=TRUE)
opts_chunk$set(tidy=TRUE, tidy.opts=list(width.cutoff=70))
```

# Introduction

The `MethylSeqR` package provides basic functions for importing and 
working with nanopore methylation sequencing data.

This vignette will walk you through a typical workflow using CH3 files created 
by Wasatch Biolabs (https://www.wasatchbiolabs.com/). These files contain results 
from a nanopore sequencing run. A CH3 file will be created for each sample within 
a run, and is meant as an intermediate existing to compress large amounts of 
data into a usable form. 

It's most helpful to store all the sample CH3 files from a run you want to 
analyze within a single directory to get started. 

# Loading Data
The first step of a workflow will be to upload data from a directory holding 
sample sequencing data in the form of CH3 files. These files are typically 
created by Wasatch Biolabs.

Once data is obtained in this format, the first step is done using the 
`make_pos_db()` function. This function directly imports a directory of CH3 
files, and creates a database in the form of a local file.

## make_pos_db()
`make_pos_db()` requires 2 arguments. First, A string representing the directory
path containing CH3 files (`ch3_files`), and second, a string representing the 
path where the positions database will be created (`ch3_db`). 

4 additional special arguments exist if needed. `chrs` allows a list representation 
of all chromosomes wanted to be included in the positional data, not limited to 
the human genome. `min_call_prob` is a numeric value representing the minimum call 
probability threshold.  Only calls with a probability greater than or equal to 
this value will be included. Default is 0.9. `min_length` is a numeric value 
representing the minimum read length. Only reads meeting or exceeding this length
will be processed. Default is 100. And `min_base_qual` is a numeric value 
representing the minimum base quality. Only reads with quality scores at or above 
this threshold will be included. Default is 10.

Here is an example:

```{r}
# Set up the file path for the test data located in inst/test_data/
ch3_files <- system.file("test_data", package = "MethylSeqR")
ch3_db <- tempfile("example_db")

# Run the function with the example data
result <- make_pos_db(ch3_files, ch3_db)
print(result)
```

While the function builds the positions table, a progress bar will appear for 
the user. The output will print the last table added to the database, all tables
currently in the database, and then print the table you just created 
(in this case, positions).

After this step, a positional data table is made, and a user can continue their 
analysis in any direction they want. They may run a quality control 
(`qc_wrapper()`), summarize positional data by regions (`summarize_regions()`) 
or windows (`summarize_windows()`). At any point in the analysis, a user can 
export a table in the database by calling `export_tables()`, or use `get_table()`
to import as a tibble into the local environment. All these functions will be 
explained below.

## summarize_regions
This function summarizes positional methylation data (from the positions table) 
based on specified regions defined in a BED, TSV, or CSV file. It performs a 
join operation between the methylation data and the regions specified in the 
annotation file, allowing for different types of joins.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")

region_bed = system.file("Islands_hg38_test.csv", package = "MethylSeqR")
# Summarize Regions using annotation table
ummarize_regions(ch3_db, region_bed)
```

There are 2 required arguments. `ch3_db` is a list containing the database file 
path. This should be a string representing a path to the database, or a valid 
"ch3_db" class object. `region_file` asks for a string representing the 
path to the BED or CSV file that contains the region annotations.`join_type` 
default argument is set to "inner", however a stringindicating the type of join 
to perform can also include "left", "right", or "full".

## summarize_windows
This function summarizes methylation data from a ch3_db database by creating 
sliding windows over the specified genomic regions. It allows for the adjustment 
of window size and step size to control the granularity of the summarization.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
 
# Calculate windows
summarize_windows(ch3_db, window_size = 100, step_size = 100)
```

There is only 1 required arguments. `ch3_db` is a list containing the database 
file path. This should be a string representing a path to the database, or a valid
"ch3_db" class object. `call_type` (string) indicates the type of data to summarize. 
Default is "positions". `window_size` (integer) specifys the size of the sliding 
window in base pairs. Default is 1000. `step_size` (integer) specifies the 
number of base pairs to step forward with each window. Default is 10. 
`overwrite` is a logical indicating whether to overwrite the existing "windows" 
table if it exists. Default is TRUE.
 
# Conclusion
In this vignette, we have walked you through the basic functions in the 
`MethylSeqR` package. This work is a work in progress and we hope to 
improve its functionality. If you have any suggestions or 
requested features, please email Jonathon Hill at jhill@byu.edu.
